### ISSUE TITLE
OMNeT++ 6.0pre8 opp_makemake builds Makefiles that "BSD make" can no longer interpret correctly

### ISSUE THREAD

**Issue by sommer** [omnetpp/omnetpp#836]
**Platform**
 - OMNeT++ version: 6.0pre8 
 - OS: macOS 10.15

**Describe the bug**

`opp_makemake` of 6.0pre8 builds Makefiles that `/usr/bin/make` on macOS does not interpret correctly.

Strictly speaking, this not a bug as running `make` from OMNeT++'s `tools/macosx/bin/make` (e.g., by sourcing OMNeT++'s `setenv` before building a simulation - as one probably should) or installing and running `gmake` works around the issue, but I don't think this was a conscious decision to enforce that only the `make` that ships with OMNeT++ can build simulations, so I'd rather ask.

**To Reproduce**

```
% git clone [INET 3.6.7]
% make
cd src && /Applications/Xcode.app/Contents/Developer/usr/bin/make
[...]
Building...
make[1]: *** No rule to make target `../out/clang-release/src/.last-copts', needed by `inet/opp_defines.h'.  Stop.
```

Other projects throw even weirder error messages ("`No rule to make target X.o, needed by X.dylib`").

This was introduced by https://github.com/omnetpp/omnetpp/commit/e41c4afde3aeb71b0d6768b5bd666b3da23918d1#diff-1982c18175bf142e3cf2d6d83a61a4afL1179

This commit changed `opp_makemake` to generate `Makefile`s using GNU extensions, which BSD make seems to have problems with. Changing `$(file >X,Y)` back to `$(shell echo "Y" >X)` restores `Makefile`s back to BSD compatibility.

**Comment by rhornig** [comment]
The change to the $(file) syntax was a tradeoff, because the previous solution broke down when CFLAGS contained a " (quote) delimited word. I was not able to figure out how to quote/escape to make it work, so I fell back to the $(file) syntax that was intended especially for this scenario. We do not test or aim to work with tools other than what we are distributing with omnet. Still if we can figure out how to do the quoting/escaping we can revert this as more compatibility is always better.

**Comment by sommer** [comment]
I can totally understand that fixing bugs with third-party build systems cannot be a priority. Still, substituting `\` with `\\` and `"` with `\"` in the `echo` seems to work on macOS and Windows and works around the need for using the `file` command. I'm linking a patch from https://github.com/omnetpp/omnetpp/pull/837

**Comment by rhornig** [comment]
I will take a second look. But as far as I remember, I've tried the trivial and not so trivial combinations and none of them worked correctly. That was the point when I decided to go with $(file). Note that the builds were always successful, but the original problem with the old method was that the dependency checking was not working and in certain cases some files were unnecessary rebuilt.

**Comment by sommer** [comment]
Thanks, I appreciate that! To the extent I could, I verified that https://github.com/omnetpp/omnetpp/pull/837 re-builds if (and only if) CFLAGS were changed, using both CFLAGS containing a broad mix of parameters containing `'`, `"`, `\`, ` `, and serial/nested combinations thereof

**Comment by ledvinap** [comment]
@rhornig : It seems that one problem with 5.6.2 `opp_makemake` was using `echo` to generate .last-ops. This added newline to `.last-ops` file and `ifneq` always triggered ... using `echo -n` seems to fix this

```
COPTS_FILE = $O/.last-copts
ifneq ("$(COPTS)","$(shell cat $(COPTS_FILE) 2>/dev/null || echo '')")
$(shell $(MKPATH) "$O" && echo -n "$(COPTS)" >$(COPTS_FILE))
endif
```

**Comment by rhornig** [comment]
After a few additional check, I have identified several features we are using in our makefiles that do not have proper workaround in older make versions. Sadly the /usr/bin/make program is also gnu make (not BSD make or whatever), but version 3.8 and it will never be updated for later versions in future (because of moving to GPL3). The GNU toolchain on macOS has been abandoned and deprecated, so we must rely on the tools we bundle with omnet. The same holds true for the bison and yacc tools, too. So we made the decision to support GNU make 4.2 or later and bundle the makefile if the given app is not available by default.

As a consequence, I'm closing this issue.
